### PR TITLE
Handle notification tap for place added

### PR DESCRIPTION
**Pull request by cp-ishita-g**
- [x] register fcm token
- [x] implement notification handler
- [x] move fcm field from ApiUserSession to ApiUser
- [x] create observer for new place added instead of http callable
- [x] manage background handler from android side to handle received fcm

https://github.com/user-attachments/assets/77f4c57e-1de0-4b85-b0d1-ab2a539c5658

### CODE REVIEW

File: app/android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -1,17 +1,51 @@
...
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.VIBRATE"/>
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
     <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION"/>
     <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
     <uses-permission android:name="android.permission.REQUEST_IGNORE_BATTERY_OPTIMIZATIONS" />
     <uses-permission android:name="android.permission.FOREGROUND_SERVICE" />
+    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED"/>
+    <uses-permission android:name="android.permission.WAKE_LOCK"/>
```
Comment (cp-radhika-s) [inline app/android/app/src/main/AndroidManifest.xml:12]: Why do we need this permission?

File: app/android/app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,17 +1,51 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.VIBRATE"/>
```
Comment (cp-radhika-s) [inline app/android/app/src/main/AndroidManifest.xml:4]: Why this permission?

File: app/android/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -1,17 +1,51 @@
...
+
+        <service android:name="com.google.firebase.messaging.FirebaseMessagingService">
+            <intent-filter>
+                <action android:name="com.google.firebase.MESSAGING_EVENT"/>
+            </intent-filter>
+        </service>
+
+        <receiver
+            android:exported="true"
+            android:name="com.google.firebase.iid.FirebaseInstanceIdReceiver"
```
Comment (cp-radhika-s) [inline app/android/app/src/main/AndroidManifest.xml:38]: What is use of this?

File: app/android/app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -1,17 +1,51 @@
...
+            android:permission="com.google.android.c2dm.permission.SEND">
+            <intent-filter>
+                <action android:name="com.google.android.c2dm.intent.RECEIVE"/>
+                <category android:name="com.canopas.yourspace"/>
+            </intent-filter>
+        </receiver>
+
+        <meta-data android:name="firebase_analytics_collection_deactivated" android:value="true" />
+        <meta-data android:name="firebase_messaging_auto_init_enabled" android:value="false" />
```
Comment (cp-radhika-s) [inline app/android/app/src/main/AndroidManifest.xml:47]: why do we need`firebase_analytics_collection_deactivated` & `firebase_messaging_auto_init_enabled`?

File: app/android/app/src/main/res/drawable/splash_inset.xml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (cp-radhika-s) [inline app/android/app/src/main/res/drawable/splash_inset.xml:1]: Why do we need this?

File: app/android/app/src/main/res/values-night/styles.xml [hunk 6]
```diff
@@ -4,7 +4,8 @@
     <style name="LaunchTheme" parent="@android:style/Theme.Black.NoTitleBar">
         <!-- Show a splash screen on the activity. Automatically removed when
              the Flutter engine draws its first frame -->
-        <item name="android:windowBackground">@drawable/launch_background</item>
+        <item name="android:windowSplashScreenBackground">#000000</item>
+        <item name="android:windowSplashScreenAnimatedIcon">@drawable/splash_inset</item>
```
Comment (cp-radhika-s) [inline app/android/app/src/main/res/values-night/styles.xml:8]: Why do we have splash screen changes here?

File: app/android/app/src/main/res/values/styles.xml [hunk 7]
```diff
@@ -4,7 +4,9 @@
     <style name="LaunchTheme" parent="@android:style/Theme.Light.NoTitleBar">
         <!-- Show a splash screen on the activity. Automatically removed when
              the Flutter engine draws its first frame -->
-        <item name="android:windowBackground">@drawable/launch_background</item>
+        <item name="android:windowSplashScreenBackground">#ffffff</item>
+        <item name="android:windowSplashScreenAnimatedIcon">@drawable/splash_inset</item>
```
Comment (cp-radhika-s) [inline app/android/app/src/main/res/values/styles.xml:8]: Remove this keep default behaviour

File: app/lib/domain/fcm/notification_handler.dart [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        break;
+      case NotificationGeofenceConst.NOTIFICATION_TYPE_GEOFENCE:
+        break;
+    }
+  }
+
+  void _handlePlaceAdded(BuildContext context, Map<String, dynamic> data) {
+    final String? spaceId = data[NotificationPlaceConst.KEY_SPACE_ID];
+    if (spaceId != null) {
+      AppRoute.placesList(spaceId).push(context);
```
Comment (cp-radhika-s) [inline app/lib/domain/fcm/notification_handler.dart:150]: Don't we need to also set this space id as the current selected space id?

File: app/lib/domain/fcm/notification_handler.dart [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:yourspace_flutter/ui/app_route.dart';
+
+const YOUR_SPACE_CHANNEL_MESSAGE = "your_space_notification_channel_messages";
+const YOUR_SPACE_CHANNEL_PLACES = "your_space_notification_channel_places";
+const YOUR_SPACE_CHANNEL_GEOFENCE = "your_space_notification_channel_geofence";
+
+const _androidChannel = AndroidNotificationChannel(
+  'notification_channel_your_space_regional',
+  'Notifications',
```
Comment (cp-radhika-s) [inline app/lib/domain/fcm/notification_handler.dart:17]: YourSpace Notification

File: app/lib/main.dart [hunk 10]
```diff
@@ -28,8 +29,15 @@ void main() async {
   }
 
   final container = await _initContainer();
+  FirebaseMessaging.onBackgroundMessage(firebaseMessagingBackgroundHandler);
+
   final userId = await _getUserIdFromPreferences();
   final isLocationPermission = await Permission.location.isGranted;
+  await Permission.notification.isDenied.then((value) {
```
Comment (cp-radhika-s) [inline app/lib/main.dart:36]: We shouldn't  ask for permission here, we have enable permission screen, where we're asking for required permission, check android for reference

**Comment by cp-ishita-g** [comment]
@kaushiksaliya please review

File: app/lib/ui/flow/home/home_screen.dart [hunk 11]
```diff
@@ -23,10 +26,34 @@ class HomeScreen extends ConsumerStatefulWidget {
...
+
+  @override
+  void initState() {
+    super.initState();
+    runPostFrame(() {
+      notificationHandler = ref.read(notificationHandlerProvider);
+      notificationHandler.init(context);
+
+      notifier = ref.watch(homeViewStateProvider.notifier);
+      notifier.getAllSpaces();
```
Comment (cp-radhika-s) [inline app/lib/ui/flow/home/home_screen.dart:39]: `notifier.getAllSpaces();` let's make this call from notifier itself

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

File: app/lib/main.dart (hunk 10 above)
Comment (cp-ishita-g) [reply app/lib/main.dart:36]: this implementation was not done by me and don't know reason why @kaushiksaliya ask for permission here

File: app/android/app/src/main/res/values/styles.xml (hunk 7 above)
Comment (cp-ishita-g) [reply app/android/app/src/main/res/values/styles.xml:8]: this changes was done by @kaushiksaliya i merge main changes in mine and reset that commit so it is showing my changes and same for above two comment

File: app/android/app/src/main/AndroidManifest.xml (hunk 4 above)
Comment (cp-ishita-g) [reply app/android/app/src/main/AndroidManifest.xml:47]: before i am not getting notification so i added this permission and also above added comment

File: app/lib/domain/fcm/notification_handler.dart (hunk 8 above)
Comment (cp-ishita-g) [reply app/lib/domain/fcm/notification_handler.dart:150]: why to make it current selected space here, user is just getting notified about that place which was added this by user, i don't think there is need to make it current space id

File: app/lib/main.dart (hunk 10 above)
Comment (kaushiksaliya) [reply app/lib/main.dart:36]: I am not adding this notification permission request. if you asking for location permission then it is required for geolocator .

File: app/android/app/src/main/res/values/styles.xml (hunk 7 above)
Comment (kaushiksaliya) [reply app/android/app/src/main/res/values/styles.xml:8]: we have to app logo one is for app icon and other is for splash with white background and also splash screen icon color is different.

File: app/android/app/src/main/AndroidManifest.xml (hunk 4 above)
Comment (cp-radhika-s) [reply app/android/app/src/main/AndroidManifest.xml]: I don't think it has anything to do with notification

File: app/lib/domain/fcm/notification_handler.dart (hunk 8 above)
Comment (cp-radhika-s) [reply app/lib/domain/fcm/notification_handler.dart:150]: here we're navigating the user to the places screen, which is deepened on selected space

File: app/android/app/src/main/AndroidManifest.xml (hunk 4 above)
Comment (cp-ishita-g) [reply app/android/app/src/main/AndroidManifest.xml]: i removed all unwanted permission, and i added this from any of the reference,

File: app/lib/domain/fcm/notification_handler.dart (hunk 8 above)
Comment (cp-ishita-g) [reply app/lib/domain/fcm/notification_handler.dart:150]: it is when we are in app, it is based on selected space id but here we are getting space id from remote notification message, so don't need to update current space id

File: data/lib/api/auth/api_user_service.dart [hunk 12]
```diff
@@ -146,6 +151,23 @@ class ApiUserService {
...
+    if (currentUser == null) return;
+    logger.d('UserService: registerDevice begin');
+
+    try {
+      final deviceToken = await FirebaseMessaging.instance.getToken() ?? "";
+      if (deviceToken.isEmpty) {
+        logger.e('UserService: registerDevice error deviceToken is null');
+        return;
+      }
+      await registerFcmToken(currentUser?.id ?? '',  deviceToken);
```
Comment (cp-radhika-s) [inline data/lib/api/auth/api_user_service.dart:164]: let's not pass an empty string, use the not null operator

Comment (cp-radhika-s) [reply data/lib/api/auth/api_user_service.dart:164]: Are we updating the session `isActive` flag when the user logout? if not then let's fix it

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

File: app/android/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (cp-radhika-s) [reply app/android/app/src/main/AndroidManifest.xml:10]: Let me know why we need this

**Review by cp-radhika-s — CHANGES_REQUESTED** [review]

File: app/android/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (cp-ishita-g) [reply app/android/app/src/main/AndroidManifest.xml:10]: are you talking about wave lock than that is added by @kaushiksaliya and receive boot completed permission error i added when i am not getting notification on android side

Comment (kaushiksaliya) [reply app/android/app/src/main/AndroidManifest.xml:10]: No, I have not added this permission.

**Comment by cp-radhika-s** [comment]
Where do we ask for notification permission?

**Comment by cp-radhika-s** [comment]
Please fix pipelines

**Comment by cp-ishita-g** [comment]
> Where do we ask for notification permission?

it was asked from enable permission view file, do we need to call again

**Review by cp-radhika-s — APPROVED** [review]

**Pull request merged**